### PR TITLE
easy-mmode-defmap, still broken

### PR DESCRIPTION
**Pull request by bitemyapp**
Do not merge.

 I wrapped it in that macro, but the stuff that needs to be an immediately-invoke-on-load isn't bound. watdo?

### CODE REVIEW

**Comment by bitemyapp** [comment]
#669

File: ghci-script-mode.el [hunk 1]
```diff
@@ -52,7 +52,9 @@
 ;;;###autoload
 (add-to-list 'auto-mode-alist '("\\.ghci\\'" . ghci-script-mode))
 
-(define-key ghci-script-mode-map (kbd "C-c C-l") 'ghci-script-mode-load)
+(easy-mmode-defmap ghci-script-mode-map
+  `("C-c C-l" . ghci-script-mode-load)
```
Comment (geraldus) [inline ghci-script-mode.el]: `(kbd "C-c C-l")`?

Comment (geraldus) [inline ghci-script-mode.el]: Oh, I suppose it's not enough. Likely it should be:

`((kbd "C-c C-l") . #'ghci-script-mode-load)`.

`#'` explanation: http://endlessparentheses.com/get-in-the-habit-of-using-sharp-quote.html

File: haskell-cabal.el [hunk 2]
```diff
@@ -116,21 +116,20 @@
...
+    ("M-n" . haskell-cabal-next-subsection)
+    ("M-p" . haskell-cabal-previous-subsection)
+    ("C-<down>" . haskell-cabal-next-subsection)
+    ("C-<up>" . haskell-cabal-previous-subsection)
+    ("C-c C-f" . haskell-cabal-find-or-create-source-file)
+    ("M-g l" . haskell-cabal-goto-library-section)
+    ("M-g e" . haskell-cabal-goto-executable-section)
+    ("M-g b" . haskell-cabal-goto-benchmark-section)
+    ("M-g t" . haskell-cabal-goto-test-suite-section))
+  "Keymap for haskell-cabal-mode")
```
Comment (geraldus) [inline haskell-cabal.el]: same here

File: haskell-debug.el [hunk 3]
```diff
@@ -68,16 +68,18 @@
...
+    ("s"   . haskell-debug/step)
+    ("t"   . haskell-debug/trace)
+    ("d"   . haskell-debug/delete)
+    ("b"   . haskell-debug/break-on-function)
+    ("a"   . haskell-debug/abandon)
+    ("c"   . haskell-debug/continue)
+    ("p"   . haskell-debug/previous)
+    ("n"   . haskell-debug/next)
+    ("RET" . haskell-debug/select))
+  "The base keymap for haskell-debug-mode")
```
Comment (geraldus) [inline haskell-debug.el]: sharp quotes here and so on

**Comment by bitemyapp** [comment]
@geraldus a *lot* of code does it that way unless you mean it addresses the self-call problem.

http://www.emacswiki.org/emacs/diffstat.el

https://github.com/haskell/haskell-mode/issues/669

**Comment by geraldus** [comment]
@bitemyapp first of all this is from your second link ``(("\C-c\C-l" . diffstat-goto-diff)`, so at least you should escape controls in `"C-c"` and etc., and I believe this is what `kbd` function for.
Secondly, [these](https://github.com/haskell/haskell-mode/pull/668/files#diff-826c9dbdf11f0d9007e1655e7169fac1R45) keybindings perfectly work. Sharp quote denoted to compiler that following symbol is defenetely is a function.

**Comment by geraldus** [comment]
Here is documentation:

> (easy-mmode-defmap M BS DOC &rest ARGS)
> Define a constant M whose value is the result of `easy-mmode-define-keymap'.

> (easy-mmode-define-keymap BS &optional NAME M ARGS)
> Return a keymap built from bindings BS.
> BS must be a list of (KEY . BINDING) where
> KEY and BINDINGS are suitable for `define-key'.

Sharp quoted symbol is suitable binding for `define-key`, so I suppose this is why it works for `easy-mmode-defmap` too.

File: haskell.el [hunk 4]
```diff
@@ -35,18 +35,16 @@
...
+(easy-mmode-defmap interactive-haskell-mode-map
+  `(("C-c C-l" . haskell-process-load-or-reload)
+    ("C-c C-t" . haskell-process-do-type)
+    ("C-c C-i" . haskell-process-do-info)
+    ("M-." . haskell-mode-jump-to-def-or-tag)
+    ("C-c C-k" . haskell-interactive-mode-clear)
+    ("C-c C-c" . haskell-process-cabal-build)
+    ("C-c C-x" . haskell-process-cabal)
+    ([?\C-c ?\C-b] . haskell-interactive-switch)
+    ([?\C-c ?\C-z] . haskell-interactive-switch))
```
Comment (geraldus) [inline haskell.el]: Also, would it be great not to mix different ways of key definition, won't it?

Comment (bitemyapp) [inline haskell.el]: @geraldus that's why I did the strawman PR.

The haskell-debug-mode bindings, for example, work fine as they are currently defined.

**Comment by gracjan** [comment]
@geraldus, @bitemyapp: We should follow what people in packages distributed inside of Emacs do, for example `smerge` has this definition:

```el
(easy-mmode-defmap smerge-basic-map
  `(("n" . smerge-next)
    ("p" . smerge-prev)
    ("r" . smerge-resolve)
    ("a" . smerge-keep-all)
    ("b" . smerge-keep-base)
    ("o" . smerge-keep-other)
    ("m" . smerge-keep-mine)
    ("E" . smerge-ediff)
    ("C" . smerge-combine-with-next)
    ("R" . smerge-refine)
    ("\C-m" . smerge-keep-current)
    ("=" . ,(make-sparse-keymap "Diff"))
    ("=<" "base-mine" . smerge-diff-base-mine)
    ("=>" "base-other" . smerge-diff-base-other)
    ("==" "mine-other" . smerge-diff-mine-other))
  "The base keymap for `smerge-mode'.")
```

So:
- [ ] no `kbd`
- [ ] no sharp-quotes
- [ ] control chars defined as `"\C-m"`

No sharp-quotes looks also strange to me but this is what people do so lets follow the crowd.

**Comment by geraldus** [comment]
@bitemyapp sorry if I was unhelpful, I'm not an Emacs Lisp expert and thought that my comments could be relevant to your problem. One other thing is looking wrong for me is this line https://github.com/haskell/haskell-mode/pull/685/files#diff-c9062b97a2be4b8029d22242507117e6R84. Travis reports error on line 20 of `ghci-script-mode.el`, but it looks spurious. Maybe line 84 is a culprit?

**Comment by bitemyapp** [comment]
@geraldus It's indeed spurious, that's me trying to get it to "fire" on mode load.

**Comment by bitemyapp** [comment]
It seems the question isn't going to be answered so I'm yanking this.

**Comment by gracjan** [comment]
If you mean this question "I wrapped it in that macro, but the stuff that needs to be an immediately-invoke-on-load isn't bound. watdo?" then I think I do not understand it.

Which parts are "immediately-invoke-on-load"?

And general comment: preferably nothing should be "immediately-invoke-on-load" as a general rule. Can we make it so it not "invoke-on-load"?

(It is better to keep this PR open, otherwise it is too easy to forget or miss it in github interface, so reopening)

File: haskell-interactive-mode.el [hunk 5]
```diff
@@ -62,27 +62,26 @@ interference with prompts that look like haskell expressions."
...
+    ("M-p" . haskell-interactive-mode-history-previous)
+    ("M-n" . haskell-interactive-mode-history-next)
+    ("C-c C-p" . haskell-interactive-mode-prompt-previous)
+    ("C-c C-n" . haskell-interactive-mode-prompt-next)
+    ("C-<up>" . haskell-interactive-mode-history-previous)
+    ("C-<down>" . haskell-interactive-mode-history-next)
+    ("TAB" . haskell-interactive-mode-tab)
+    ("<C-S-backspace>" . haskell-interactive-mode-kill-whole-line))
   "Interactive Haskell mode map.")
+(haskell-interactive-mode-map)
```
Comment (geraldus) [inline haskell-interactive-mode.el]: This line seems to be there by mistake for me.

**Comment by geraldus** [comment]
@bitemyapp oh, why you closed this? You did a great work!
And as for me I also didn't catch your question. Can you clarify a bit or rephrase?

**Comment by bitemyapp** [comment]
@geraldus I opened it to ask one question. I wasn't looking for feedback on anything else except for that one question.

@gracjan could we add a question or strawman issue/PR tag? It's more efficient for me to just PR stuff when I have a question. I don't have time to do Q&A in IRC, so it has to be asynchronous.

**Comment by gracjan** [comment]
Okay, good. Can you point me to the stuff that is 'immediately-invoke-on-load'?

**Comment by bitemyapp** [comment]
@gracjan almost definitely using the wrong terminology, but: https://github.com/haskell/haskell-mode/pull/685/files#diff-fa6c31455261917e156b4ecb38c4e5ebL71

I'm guessing I need to add a reference to the var to fire off the key bindings somewhere/somehow?

**Comment by gracjan** [comment]
The statement:

```el
(define-derived-mode haskell-debug-mode ...)
```

Defined new mode and also a variable called `haskell-debug-mode-map`. Your part fills this var, which is good. This mode-map will be working when `haskell-debug-mode` is enabled. You can artificially enable it for some buffer by issuing `M-x haskell-debug-mode`, functions will fail but at least you will see if keybidings work.

**Comment by bitemyapp** [comment]
@gracjan haskell-debug-mode presently works, I think the interactive stuff did not.

**Comment by bitemyapp** [comment]
Migrating this to a topic branch per `CONTRIBUTING.md`

**Pull request closed without merging**